kernel_tuner/cupy.py

```diff
@@ -59,8 +59,7 @@ def __init__(self, device=0, iterations=7, compiler_options=None, observers=None
             raise ImportError("Error: cupy not installed, please install e.g. using 'pip install cupy-cuda111', please check https://github.com/cupy/cupy.")
 
         #select device
-        self.dev = dev = cp.cuda.Device(device)
-        dev.use()
+        self.dev = dev = cp.cuda.Device(device).__enter__()
```

Comment:
I did not know about the __enter__ method, as it is not really documented :)
If I understand this correctly it is the method called when you use the **with** statement. Is this permanent?

Reply:
Yes indeed! The enter and exit methods are what is used to enter and exit a context manager with a *with* statement. We call it here instead of with a with-statement because the device functions (up to the runners even) are also context managers themselves.

We introduced the concept of context managers (with help from Willem Jan) to solve a bug with PyCuda that occurred when users make two calls to Kernel Tuner consecutively. The context of the first call was destroyed at an arbitrary moment, when the garbage collector finally cleaned up the already actively deleted PyCuda context from the first call. And because PyCuda maintains a stack of contexts the newly created 2nd context was popped instead of the old 1st context. Using context managers here is a way to force clean up of GPU resources exactly when we want to.